File: packages/shared/src/types/chantier.ts
import { BaseEntity, UUID } from './common';

export type ChantierStatut =
  | 'lead'
  | 'visite_planifiee'
  | 'devis_envoye'
  | 'accepte'
  | 'planifie'
  | 'en_cours'
  | 'termine'
  | 'facture'
  | 'annule';

export type TypePrestation =
  | 'paysagisme'
  | 'entretien'
  | 'elagage'
  | 'abattage'
  | 'tonte'
  | 'taille'
  | 'autre';

export interface Chantier extends BaseEntity {
  clientId: UUID;
  // Localisation
  adresse: string;
  codePostal: string;
  ville: string;
  coordonnees?: {
    latitude: number;
    longitude: number;
  };
  // Details
  typePrestation: TypePrestation[];
  description: string;
  surface?: number;
  // Planning
  statut: ChantierStatut;
  dateVisite?: Date;
  dateDebut?: Date;
  dateFin?: Date;
  // Responsables
  responsableId?: UUID;
  equipeIds?: UUID[];
  // Metadata
  notes?: string;
  photos?: string[];
}

export interface CreateChantierDto {
  clientId: UUID;
  adresse: string;
  codePostal: string;
  ville: string;
  coordonnees?: {
    latitude: number;
    longitude: number;
  };
  typePrestation: TypePrestation[];
  description: string;
  surface?: number;
  dateVisite?: Date;
  notes?: string;
}

export interface UpdateChantierDto extends Partial<CreateChantierDto> {
  statut?: ChantierStatut;
  dateDebut?: Date;
  dateFin?: Date;
  responsableId?: UUID;
  equipeIds?: UUID[];
}

export interface ChantierFilters {
  clientId?: UUID;
  statut?: ChantierStatut | ChantierStatut[];
  typePrestation?: TypePrestation;
  ville?: string;
  responsableId?: UUID;
  dateDebutFrom?: Date;
  dateDebutTo?: Date;
}

File: apps/mobile/lib/domain/models/chantier.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../enums/chantier_statut.dart';
import '../enums/type_prestation.dart';

part 'chantier.freezed.dart';
part 'chantier.g.dart';

@freezed
abstract class Chantier with _$Chantier {
  const factory Chantier({
    required String id,
    required String clientId,
    required String adresse,
    required String codePostal,
    required String ville,
    double? latitude,
    double? longitude,
    @Default([]) List<TypePrestation> typePrestation,
    required String description,
    double? surface,
    @Default(ChantierStatut.lead) ChantierStatut statut,
    DateTime? dateVisite,
    DateTime? dateDebut,
    DateTime? dateFin,
    String? responsableId,
    String? notes,
    @Default([]) List<String> photos,
    DateTime? deletedAt,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _Chantier;

  factory Chantier.fromJson(Map<String, dynamic> json) =>
      _$ChantierFromJson(json);
}
